### PR TITLE
Fix: avoid package name collision with inPackage (#291)

### LINKED ISSUE #291: Invalid mock generated with the same package names
Using either 1.1.2 or 2.0.0-alpha.13 release builds of mockery we have the following issue

`cloudformation/interface.go`
```go
package cloudformation

import "github.com/aws/aws-sdk-go/service/cloudformation"

type invalid interface {
	String(*cloudformation.CreateStackInput) string
}
```

`main.go`
```go
package main

func main() {}
```

generates a mock like like the following when using `mockery -all -testonly -inpkg`
```go
// Code generated by mockery v1.1.2. DO NOT EDIT.

package cloudformation

import mock "github.com/stretchr/testify/mock"

// mockInvalid is an autogenerated mock type for the invalid type
type mockInvalid struct {
	mock.Mock
}

// String provides a mock function with given fields: _a0
func (_m *mockInvalid) String(_a0 *cloudformation.CreateStackInput) string {
	ret := _m.Called(_a0)

	var r0 string
	if rf, ok := ret.Get(0).(func(*cloudformation.CreateStackInput) string); ok {
		r0 = rf(_a0)
	} else {
		r0 = ret.Get(0).(string)
	}

	return r0
}
```

You can see the import for cloudformation is missing and in turn the go reports undefined: cloudformation in the mock file.

It seems like when the cloudformation directory is at the root the issue cannot be re-created. When mockery is built from source either via `go get` or cloning and `go install` this issue seems to go away and the import is generated properly. As installing is not supported using `go get` anymore our mocks are no longer generating properly.
```go
package cloudformation

import (
	"github.com/aws/aws-sdk-go/service/cloudformation"
	mock "github.com/stretchr/testify/mock"
)
```

### PR DESCRIPTION
**Pull request by i-sevostyanov**
Fixes #291.

Tired of waiting, so I copied everything from https://github.com/vektra/mockery/pull/402, sorry.

### CODE REVIEW

**Comment by fterrag** [comment]
Can we get this PR reviewed, please? Thanks!

**Comment by peppelan** [comment]
same here, would be great, thankyousomuchplease

**Comment by LandonTClipp** [comment]
@strideynet sorry for the delay, life gets in the way sometimes, I'm trying to work through a backlog of stuff. Can you write in the description more succinctly what the solution to the original problem is? This is just to aid me so I can know better what I'm looking at.

**Comment by LandonTClipp** [comment]
@i-sevostyanov  @strideynet sorry for the delay, life gets in the way sometimes, I'm trying to work through a backlog of stuff. Can you write in the description more succinctly what the solution to the original problem is? This is just to aid me so I can know better what I'm looking at.

**Comment by i-sevostyanov** [comment]
I think this issue describes the problem in detail: [Invalid mock generated with the same package names](https://github.com/vektra/mockery/issues/291) 🤓

**Comment by serivkos** [comment]
@LandonTClipp I would appreciate if you could look over this request and merge it with master. I’ll be waiting for that. Thanks

**Comment by fterrag** [comment]
We're currently working around this with some `sed` magic - looking forward to a fix!

File: pkg/generator.go [hunk 1]
```diff
@@ -104,7 +104,10 @@ func (g *Generator) addPackageImportWithName(ctx context.Context, path, name str
 
 func (g *Generator) getNonConflictingName(path, name string) string {
 	if !g.importNameExists(name) {
-		return name
+		// do not allow imports with the same name as the package when inPackage
+		if !g.InPackage || g.iface.Pkg.Name() != name {
+			return name
```
Comment (LandonTClipp) [inline pkg/generator.go:109]: Let's collapse this if statement into one

```go
if !g.importNameExists(name) && (!g.InPackage || g.iface.Pkg.Name() != name)
```

File: pkg/generator.go [hunk 2]
```diff
@@ -121,7 +124,10 @@ func (g *Generator) getNonConflictingName(path, name string) string {
 	for i := 1; i <= numDirectories; i++ {
 		prospectiveName = strings.Join(cleanedDirectories[numDirectories-i:], "")
 		if !g.importNameExists(prospectiveName) {
-			return prospectiveName
+			// do not allow imports with the same name as the package when inPackage
```
Comment (LandonTClipp) [inline pkg/generator.go:127]: Same, let's collapse into one if statement.

File: pkg/fixtures/example_project/foo/collision.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package foo
+
+import "github.com/vektra/mockery/v2/pkg/fixtures/example_project/bar/foo"
+
+type Collision interface {
```
Comment (LandonTClipp) [inline pkg/fixtures/example_project/foo/collision.go:5]: Maybe add a comment in this describing what the intended situation here is (obviously, package and import having the same name).

**Review by LandonTClipp — CHANGES_REQUESTED** [review]
Thanks for the PR, just a few minor comments and we can get this merged.

File: pkg/generator.go (hunk 1 above)
Comment (i-sevostyanov) [reply pkg/generator.go]: Done

File: pkg/generator.go (hunk 2 above)
Comment (i-sevostyanov) [reply pkg/generator.go:125]: Done

File: pkg/fixtures/example_project/foo/collision.go (hunk 3 above)
Comment (i-sevostyanov) [reply pkg/fixtures/example_project/foo/collision.go]: I renamed the interface, so now it has a more explicit name

Comment (LandonTClipp) [reply pkg/fixtures/example_project/foo/collision.go]: Thanks @i-sevostyanov , tests seem to be failing unfortunately. I'll let you take care of that.

Comment (i-sevostyanov) [reply pkg/fixtures/example_project/foo/collision.go]: Oh, sorry for that. Fixed 🤓

**Review by LandonTClipp — APPROVED** [review]
Looks good!

**Comment by LandonTClipp** [comment]
@i-sevostyanov sorry one more thing :) the coverage report is showing some missing coverage. 

https://app.codecov.io/gh/vektra/mockery/compare/422/changes

Could you add tests for the cases that are missing? PRs generally should be increasing coverage, as a matter of principle.

**Comment by i-sevostyanov** [comment]
Done 🤓

**Pull request merged**